src/ray/raylet/node_manager.cc

```diff
@@ -931,65 +853,75 @@ void NodeManager::ProcessGetTaskMessage(
+
   // Call task dispatch to assign work to the new worker.
   DispatchTasks(local_queues_.GetReadyTasksWithResources());
 }
 
-void NodeManager::ProcessDisconnectClientMessage(
-    const std::shared_ptr<LocalClientConnection> &client, bool intentional_disconnect) {
-  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(client);
+/// Handle a `DisconnectClient` request.
+void NodeManager::HandleDisconnectClientRequest(
+    const rpc::DisconnectClientRequest &request, rpc::DisconnectClientReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  const WorkerID worker_id = WorkerID::FromBinary(request.worker_id());
+  bool intentional_disconnect = request.intentional();
+  RAY_LOG(DEBUG) << "Handle disconnect client request for worker " << worker_id;
+
+  ProcessDisconnectClientMessage(worker_id, intentional_disconnect);
+  send_reply_callback(Status::OK(), nullptr, nullptr);
+}
+
+void NodeManager::ProcessDisconnectClientMessage(const WorkerID &worker_id,
+                                                 bool intentional_disconnect) {
+  std::shared_ptr<Worker> worker = worker_pool_.GetRegisteredWorker(worker_id);
   bool is_worker = false, is_driver = false;
   if (worker) {
     // The client is a worker.
     is_worker = true;
   } else {
-    worker = worker_pool_.GetRegisteredDriver(client);
+    worker = worker_pool_.GetRegisteredDriver(worker_id);
     if (worker) {
       // The client is a driver.
       is_driver = true;
     } else {
       RAY_LOG(INFO) << "Ignoring client disconnect because the client has already "
                     << "been disconnected.";
+      return;
     }
   }
   RAY_CHECK(!(is_worker && is_driver));
 
   // If the client has any blocked tasks, mark them as unblocked. In
   // particular, we are no longer waiting for their dependencies.
   if (worker) {
-    if (is_worker && worker->IsDead()) {
-      // Don't need to unblock the client if it's a worker and is already dead.
-      // Because in this case, its task is already cleaned up.
+    if (is_worker && worker->IsKilling()) {
+      // Don't need to unblock the client if it's a worker and have sent kill signal to
+      // it. Because in this case, its task is already cleaned up.
       RAY_LOG(DEBUG) << "Skip unblocking worker because it's already dead.";
     } else {
       while (!worker->GetBlockedTaskIds().empty()) {
         // NOTE(swang): HandleTaskUnblocked will modify the worker, so it is
         // not safe to pass in the iterator directly.
         const TaskID task_id = *worker->GetBlockedTaskIds().begin();
-        HandleTaskUnblocked(client, task_id);
+        HandleTaskUnblocked(worker_id, task_id);
       }
     }
   }
 
   if (is_worker) {
-    // The client is a worker.
-    if (worker->IsDead()) {
-      // If the worker was killed by us because the driver exited,
-      // treat it as intentionally disconnected.
-      intentional_disconnect = true;
```

Comment:
why this is removed?

Reply:
I think it's not used, because we handle the process disconnect in the heartbeat function and set it as arguments outside this func.